test_conformance/spirv_new/test_decorate.cpp

```diff
@@ -370,60 +446,77 @@ int test_fp_rounding(cl_device_id deviceID,
     return 0;
 }
 
-template<typename Ti, typename To>
-inline To round_to_zero(Ti in)
+template <typename T> static inline double to_double(T in) { return in; }
+
+template <> inline double to_double(cl_half in) { return cl_half_to_float(in); }
+
+template <typename Ti, typename To> static inline To round_to_zero(Ti in)
 {
-    To out = (To)(in);
-    return out;
+    return (To)to_double(in);
 }
 
-template<typename T>
-int sign(T val)
+template <typename T> static inline int sign(T val)
 {
     if (val < 0) return -1;
     if (val > 0) return 1;
     return 0;
 }
 
-template<typename Ti, typename To>
-inline To round_to_even(Ti in)
+template <typename Ti, typename To> static inline To round_to_even(Ti in)
 {
-    // https://en.wikipedia.org/wiki/Rounding#Round_half_to_even
-    return std::floor(in + 0.5) - 1 + std::abs(sign(reference_remainderl((long double)in, 2) - 0.5));
+    double din = to_double(in);
+    return std::floor(din + 0.5) - 1
+        + std::abs(sign(reference_remainderl((long double)din, 2) - 0.5));
 }
 
-template<typename Ti, typename To>
-inline To round_to_posinf(Ti in)
+template <typename Ti, typename To> static inline To round_to_posinf(Ti in)
 {
-    To out = std::ceil(in);
-    return out;
+    return std::ceil(to_double(in));
 }
 
-template<typename Ti, typename To>
-inline To round_to_neginf(Ti in)
+template <typename Ti, typename To> static inline To round_to_neginf(Ti in)
+{
+    return std::floor(to_double(in));
+}
+
+template <typename Ti, typename To>
+static inline Ti generate_fprounding_input(RandomSeed &seed)
 {
-    To out = std::floor(in);
-    return out;
+    constexpr auto minVal = std::numeric_limits<To>::min() / 2;
+    constexpr auto maxVal = std::numeric_limits<To>::max() / 2;
```

Comment:
This results in values not representable in `cl_float`, which triggers a warning (promoted to error) when building with Clang, and potential silent unwanted behavior (inputs outside of `[minVal, maxVal]`) when building with GCC.
```
/home/harald/OpenCL-CTS/test_conformance/spirv_new/test_decorate.cpp:494:42: error: implicit conversion from 'const int' to 'float' changes value from 1073741823 to 1073741824 [-Werror,-Wimplicit-const-int-float-conversion]
    return genrandReal_range<Ti>(minVal, maxVal, seed);
           ~~~~~~~~~~~~~~~~~             ^~~~~~
/home/harald/OpenCL-CTS/test_conformance/spirv_new/test_decorate.cpp:521:1: note: in instantiation of function template specialization 'generate_fprounding_input<float, int>' requested here
TEST_SPIRV_FP_ROUNDING_DECORATE(rte, round_to_even, float, int);
^
/home/harald/OpenCL-CTS/test_conformance/spirv_new/test_decorate.cpp:509:21: note: expanded from macro 'TEST_SPIRV_FP_ROUNDING_DECORATE'
            in[i] = generate_fprounding_input<clTi, clTo>(seed);               \
                    ^
```

Reply:
I've reproduced this issue.  Working on a fix.

Reply:
I think the right thing to do here may be to simply silence the warning by inserting explicit casts.  Here's why:

This function is generating input data for tests that round from a floating-point value to an integer value.  The `minVal` and the `maxVal` are used to determine the range of input data that is tested, and is determined by the integer type we are rounding to.

Since we're dividing the minimum value and the maximum value by two we don't necessarily test the entire range of integer data, but we do test a good portion of it.  This also means, though, if we happen to test a slightly smaller range or a slightly larger range because the conversion from the integer minimum or maximum value to the floating-point minimum or maximum value is inexact, that's fine, and nothing to worry about.

I'll make a suggestion in a different review comment for review.

Reply:
" This also means ... nothing to worry about." Thanks for the explanation, that seems like it makes sense.

Reply:
Great, I've committed my suggestion and I think this is good-to-go now, just needs a review approval.